phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/CurrencyManager.kt

```diff
@@ -410,78 +489,176 @@ class CurrencyManager(
+        } catch (e: Exception) {
+            log.error { "Failed to get http response from blockchain.info: $e" }
+            null
+        }
+        val parsedResponse: BlockchainInfoResponse? = httpResponse?.let {
+            try {
+                json.decodeFromString<BlockchainInfoResponse>(it.bodyAsText())
+            } catch (e: Exception) {
+                log.error { "Failed to parse json response from blockchain.info: $e" }
+                null
+            }
+        }
 
-        if (fiatCurrencies.isEmpty()) {
-            return
-        } else {
-            log.info { "fetching ${fiatCurrencies.size} exchange rate(s) from coindesk.com" }
-            addRefreshTargets(fiatCurrencies)
+        val timestampMillis = Clock.System.now().toEpochMilliseconds()
+        val fetchedRates: List<ExchangeRate> = parsedResponse?.let {
+            targets.mapNotNull { fiatCurrency ->
+                parsedResponse[fiatCurrency.name]?.let { priceObject ->
+                    ExchangeRate.BitcoinPriceRate(
+                        fiatCurrency = fiatCurrency,
+                        price = priceObject.last,
+                        source = "blockchain.info",
+                        timestampMillis = timestampMillis,
+                    )
+                }
+            }
+        } ?: listOf()
+
+        return fetchedRates
+    }
+
+    private suspend fun fetchFromCoinbase(
+        targets: Set<FiatCurrency>
+    ): List<ExchangeRate> {
+
+        val httpResponse: HttpResponse? = try {
+            httpClient.get(urlString = "https://api.coinbase.com/v2/exchange-rates?currency=USD")
+        } catch (e: Exception) {
+            log.error { "Failed to get http response from api.coinbase.com: $e" }
+            null
         }
+        val parsedResponse: CoinbaseResponse? = httpResponse?.let {
+            try {
+                json.decodeFromString<CoinbaseResponse>(it.bodyAsText())
+            } catch (e: Exception) {
+                log.error { "Failed to parse json response from api.coinbase.com: $e" }
+                null
+            }
+        }
+
+        val timestampMillis = Clock.System.now().toEpochMilliseconds()
+        val fetchedRates: List<ExchangeRate> = parsedResponse?.let {
+            targets.mapNotNull { fiatCurrency ->
+                parsedResponse.data.rates[fiatCurrency.name]?.let { valueAsString ->
+                    valueAsString.toDoubleOrNull()?.let { valueAsDouble ->
+                        ExchangeRate.UsdPriceRate(
+                            fiatCurrency = fiatCurrency,
+                            price = valueAsDouble,
+                            source = "coinbase.com",
+                            timestampMillis = timestampMillis
+                        )
+                    }
+                }
+            }
+        } ?: listOf()
+
+        return fetchedRates
+    }
+
+    private suspend fun fetchFromCoinDesk(
```

Comment:
There's some advanced performance logic in this method that was added because the Coindesk API does not provide a single point of entry to fetch everything in one request and we had to chain ~146 requests sequentially. But that's not the case anymore since now we only use Coindesk for 7 currencies, so maybe can remove this code? Or at least update the comment to show that it's not really a concern anymore.

I'm also wondering if the new query to Coinbase (which fetches a large-ish JSON) could have any impacts on the iOS notification-service extension? If I recall correctly it was quite sensitive to performances issues.

Reply:
> I'm also wondering if the new query [...] could have any impacts on the iOS notification-service extension?

That's a good question. I will investigate.

Reply:
> Or at least update the comment to show that it's not really a concern anymore.

Done in 1ccb122. I still prefer to fetch all the Url's first, and write to the database after, since it was overall faster in previous tests.